File: frontend/src/api/questionBank.ts
import { requestJson } from './client'
import type {
  IsoDateTime,
  QuestionDifficulty,
  QuestionRevisionStatus,
  UUID,
} from './questionEditor'

export type QuestionBankSort = 'updated_desc' | 'created_desc'

export type QuestionBankListQuery = {
  q?: string
  question_type_id?: UUID
  status?: QuestionRevisionStatus
  difficulty?: QuestionDifficulty
  purpose_id?: UUID
  page?: number
  page_size?: number
  sort?: QuestionBankSort
}

export type QuestionBankQuestionTypeRead = {
  id: UUID
  name: string
  display_name: string
}

export type QuestionBankPrimaryTopicRead = {
  id: UUID
  name: string
  display_name: string
}

export type QuestionBankItemRead = {
  question_family_id: UUID
  question_form_id: UUID
  revision_id: UUID
  revision_number: number
  status: QuestionRevisionStatus
  is_current_approved: boolean
  question_type: QuestionBankQuestionTypeRead
  difficulty: QuestionDifficulty | null
  primary_topic: QuestionBankPrimaryTopicRead | null
  block_count: number
  text_preview: string | null
  updated_at: IsoDateTime
}

export type QuestionBankPageRead = {
  items: QuestionBankItemRead[]
  page: number
  page_size: number
  total: number
  total_pages: number
}

function buildQuestionBankQueryString(query: QuestionBankListQuery): string {
  const parameters = new URLSearchParams()

  if (query.q !== undefined) parameters.set('q', query.q)
  if (query.question_type_id !== undefined) {
    parameters.set('question_type_id', query.question_type_id)
  }
  if (query.status !== undefined) parameters.set('status', query.status)
  if (query.difficulty !== undefined) {
    parameters.set('difficulty', query.difficulty)
  }
  if (query.purpose_id !== undefined) {
    parameters.set('purpose_id', query.purpose_id)
  }
  if (query.page !== undefined) parameters.set('page', String(query.page))
  if (query.page_size !== undefined) {
    parameters.set('page_size', String(query.page_size))
  }
  if (query.sort !== undefined) parameters.set('sort', query.sort)

  const serialized = parameters.toString()
  return serialized ? `?${serialized}` : ''
}

export function getQuestionBankQuestions(
  accessToken: string,
  query: QuestionBankListQuery = {},
): Promise<QuestionBankPageRead> {
  const queryString = buildQuestionBankQueryString(query)

  return requestJson<QuestionBankPageRead>(
    `/api/v1/question-bank/questions${queryString}`,
    {
      method: 'GET',
      headers: {
        Authorization: `Bearer ${accessToken}`,
      },
    },
  )
}

File: backend/app/schemas/question_bank.py
from __future__ import annotations

import uuid
from datetime import datetime
from enum import Enum

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.core.enums import QuestionDifficulty, QuestionRevisionStatus


class StrictQuestionBankSchema(BaseModel):
    model_config = ConfigDict(extra="forbid")


class QuestionBankSort(str, Enum):
    UPDATED_DESC = "updated_desc"
    CREATED_DESC = "created_desc"


class QuestionBankListQuery(StrictQuestionBankSchema):
    q: str | None = Field(default=None, max_length=200)
    question_type_id: uuid.UUID | None = None
    status: QuestionRevisionStatus | None = None
    difficulty: QuestionDifficulty | None = None
    purpose_id: uuid.UUID | None = None
    source_id: uuid.UUID | None = None
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=25, ge=1, le=100)
    sort: QuestionBankSort = QuestionBankSort.UPDATED_DESC

    @field_validator("q", mode="before")
    @classmethod
    def normalize_search(cls, value: object) -> object:
        if value is None:
            return None
        if not isinstance(value, str):
            return value
        normalized = value.strip()
        return normalized or None


class QuestionBankQuestionTypeRead(StrictQuestionBankSchema):
    id: uuid.UUID
    name: str
    display_name: str


class QuestionBankPrimaryTopicRead(StrictQuestionBankSchema):
    id: uuid.UUID
    name: str
    display_name: str


class QuestionBankSourceRead(StrictQuestionBankSchema):
    id: uuid.UUID
    name: str
    display_name: str
    detail: str | None


class QuestionBankItemRead(StrictQuestionBankSchema):
    question_family_id: uuid.UUID
    question_form_id: uuid.UUID
    revision_id: uuid.UUID
    revision_number: int
    status: QuestionRevisionStatus
    is_current_approved: bool
    question_type: QuestionBankQuestionTypeRead
    difficulty: QuestionDifficulty | None
    primary_topic: QuestionBankPrimaryTopicRead | None
    source: QuestionBankSourceRead | None
    block_count: int = Field(ge=0)
    text_preview: str | None
    updated_at: datetime


class QuestionBankPageRead(StrictQuestionBankSchema):
    items: list[QuestionBankItemRead]
    page: int = Field(ge=1)
    page_size: int = Field(ge=1, le=100)
    total: int = Field(ge=0)
    total_pages: int = Field(ge=0)
